zebra-chain/src/amount.rs

```diff
@@ -326,12 +326,14 @@ where
     }
 }
 
-impl std::ops::Neg for Amount<NegativeAllowed> {
-    type Output = Self;
+impl<C> std::ops::Neg for Amount<C>
+where
+    C: Constraint,
+{
+    type Output = Result<Amount<C>>;
```

Comment:
Can we return `Amount<NegativeAllowed>` here, to avoid the `Result`?

That way, we won't need to change all the other uses of `Neg`.

Reply:
I think that if we do that we end up the same as we started, this is that we have to do 2 `constrain()` calls. One to pass from `C` to `NegativeAllowed` and another one for `NonNegative`.

Reply:
### My Suggestion

I'm suggesting that `neg` always converts from `C` to `NegativeAllowed`.

This conversion will never cause an error, because the negated value always fits in the range of `NegativeAllowed`.

Here's how we set up the trait to always return `Amount<NegativeAllowed>`:
```rust
impl<C> std::ops::Neg for Amount<C>
where
    C: Constraint,
{
    type Output = Amount<NegativeAllowed>;
    ....
}
```

Since the return value is always `Amount<NegativeAllowed>`, we'll only need to `constrain::<NonNegative>` in `remaining_transaction_value`.

### Current Code

If we stay with `fn neg(self: Amount<C>) -> Result<Amount<C>>`, it's very easy to do:
```rust
let value: Amount<NonNegative> = ... ;
(-value)?
```
which always returns an error, unless value is zero. The code looks ok, but it doesn't do what we want.

But what we really want is:
```rust
let value: Amount<NonNegative> = ... ;
-(value.constrain::<NegativeAllowed>()).expect("negating a NonNegative makes a valid NegativeAllowed")
```
which is what `fn neg(self: Amount<C>) -> Amount<NegativeAllowed>` does anyway. (But without the extra `Result` and `expect` in the calling code.)

Reply:
Thank you for the detailed response, i made this change in https://github.com/ZcashFoundation/zebra/pull/2569/commits/9f0d2b7defdfc76a5edffeec7055d75e40cf896e